### PR TITLE
Fix circular dependencies that pollutes version bump

### PR DESCRIPTION
**Pull request by fcollonval**
## References

Fix https://github.com/jupyterlab/jupyterlab/issues/11630
Fix https://github.com/jupyterlab/jupyterlab/issues/5821

We have 2 circular dependencies to break. The major trouble being testutils. We cannot depend on test utils for all packages. But as it is published, we probably needs to keep publishing it although the best approach would be to internalize the utils in each package => This is done by adding a `src/testutils.ts` that is not exposed through the `src/index.ts`.

```
2022-04-29T13:17:46.6049359Z lerna WARN ECYCLE @jupyterlab/application -> @jupyterlab/rendermime -> @jupyterlab/markedparser-extension -> @jupyterlab/application
2022-04-29T13:17:46.6050993Z lerna WARN ECYCLE @jupyterlab/apputils -> @jupyterlab/testutils -> @jupyterlab/notebook -> @jupyterlab/cells -> @jupyterlab/attachments -> (nested cycle: @jupyterlab/application -> @jupyterlab/rendermime -> @jupyterlab/markedparser-extension -> @jupyterlab/application) -> @jupyterlab/apputils
```

lerna does not ignore circular dependencies that come from dev dependencies. So when bumping the version it is spreading to far more packages than needed.

## Code changes

Create a new package `@jupyterlab/testing` that contains the configuration for babel and jest as well as all helpers for testing `@jupyterlab/services`.

> `@jupyterlab/coreutils` is the only package not using that new package for its tests.

The helpers in `@jupyterlab/testutils` have been split between the packages in a `testutils.ts` file (or folder) that is not directly exported by their `index`.

Some tests have chicken/egg trouble (packages codeeditor, completer and rendermime are using their downstream packages for unit tests :disappointed:) -> the quick solution done here where to move those test files in metapackage.

- [x] TBC The package `@jupyterlab/testutils` exports all previous helpers. But they main not be gathered under the same namespace.

## User-facing changes

N/A

## Backwards-incompatible changes

Hopefully few

### CODE REVIEW

**Comment by fcollonval** [comment]
Current top circular loop detected by lerna:

`lerna WARN ECYCLE @jupyterlab/apputils -> @jupyterlab/services -> @jupyterlab/nbformat -> @jupyterlab/testutils -> @jupyterlab/notebook -> @jupyterlab/cells -> @jupyterlab/apputils`

**Comment by jtpio** [comment]
> * TBC The package `@jupyterlab/testutils` exports all previous helpers. But they main not be gathered under the same namespace.

If there is a namespace change maybe we can add it to the migration guide: https://jupyterlab.readthedocs.io/en/latest/extension/extension_migration.html#api-breaking-changes

**Comment by jtpio** [comment]
Thanks @fcollonval for your patience :+1: 

It would have been nice to get a green CI on this one since it's quite of a big change, but not sure how feasible it will be in the short term. For reference these failures are tracked in:
- https://github.com/jupyterlab/jupyterlab/issues/13460
- https://github.com/jupyterlab/jupyterlab/issues/13587

**Comment by jtpio** [comment]
> It would have been nice to get a green CI on this one since it's quite of a big change

Will have a look today, hoping we get it in despite the failures.

**Comment by fcollonval** [comment]
The weird thing is that the flakyness is not coming from the _historically_ known flaky tests...
For the example, I think it is related to Jupyter Server v2 or some major upstream Jupyter packages releases

**Comment by jtpio** [comment]
> I think it is related to Jupyter Server v2 or some major upstream Jupyter packages releases

That was also my guess although not verified yet.

File: docs/source/extension/extension_migration.rst [hunk 1]
```diff
@@ -145,6 +145,21 @@ bumped their major version (following semver convention). We want to point out p
...
+Testing with Jest
+^^^^^^^^^^^^^^^^^
+
+Jest has been updated to 29.2.0 (and *ts-jest* to 29.0.0). And therefore the jest configuration provided by
+``@jupyterlab/testutils`` is compatible for that version. In particular:
+
+- The unmaintained reporter ``jest-summary-reporter`` has been replaced by the new default ``github-actions`` reporter.
+- The helper ``flakyIt`` has been removed. You can use the new `jest.retryTimes <https://jestjs.io/docs/jest-object#jestretrytimesnumretries-options>`_ instead.
+
+With v4, we fixed circular dependencies due to the testutils package. So it is now only a facade to export
```
Comment (jtpio) [inline docs/source/extension/extension_migration.rst:157]: Maybe we can make that one explicit here:

```suggestion
With JupyterLab 4, we fixed circular dependencies due to the testutils package. So it is now only a facade to export
```

**Comment by fcollonval** [comment]
@jtpio what should we do about the test timing out?

`jupyter.services - Comm › IComm › #onClose › should be called when the server side closes: packages/services/test/kernel/comm.spec.ts#L259`

I don't have any trouble locally. Should we skip it?

Pinging @afshin do you have some fresh ideas to try to fix/improve that test?

**Comment by jtpio** [comment]
> `jupyter.services - Comm › IComm › #onClose › should be called when the server side closes: packages/services/test/kernel/comm.spec.ts#L259`

Isn't it still relevant since it also happens on master? https://github.com/jupyterlab/jupyterlab/commit/f15709ff42dcde127fef1ccedc1ae6a078413d5e

https://github.com/jupyterlab/jupyterlab/actions/runs/3711556689/jobs/6292115706

image

Could it be related to the `ipykernel` / `comm` split upstream?

**Comment by fcollonval** [comment]
> Could it be related to the ipykernel / comm split upstream?

Thanks for the suggestion. Indeed it may be the case because after upgrading ipykernel locally, that test if failing locally too.

Pinging @martinRenou

**Comment by fcollonval** [comment]
Lately `js-services` is working again - looks like https://github.com/ipython/ipykernel/pull/1056 fixed the issue.

I restarted the job to check if this is the case here.

**Comment by fcollonval** [comment]
Ok indeed this fixes it. Big thanks to @blink1073 :1st_place_medal: :rocket:

**Comment by jtpio** [comment]
> This is done by adding a src/testutils.ts that is not exposed through the src/index.ts.

This might look a bit odd an unusual at first since this was not really a pattern used across the lab code base until now. But yeah I am also not sure whether there would be another way to fix the circular dependency without doing something like this.

File: packages/filebrowser/test/listing.spec.ts [hunk 2]
```diff
@@ -1,22 +1,27 @@
...
   model: new FilterFileBrowserModel({
     manager: new DocumentManager({
       registry: new DocumentRegistry(),
-      opener: new Mock.DocumentWidgetOpenerMock(),
-      manager: new Mock.ServiceManagerMock()
+      opener: {
+        open: () => {
+          /* noop */
+        }
+      } as any,
```
[the remark this answers is not in the document]
Comment (fcollonval) [reply packages/filebrowser/test/listing.spec.ts:23]: A miss in the many rebase

**Pull request merged**